gapir/cc/in_memory_resource_cache.cpp

```diff
@@ -23,174 +23,237 @@
+           resRecord->second.first.getSize());
   }
 
-  mHead = last;  // Move head to the space.
-  mBufferSize = newSize;
+  // Update the bookkeeping for LRU to reflect this access.
+  auto newID = mIDGenerator++;
+  mResourceIndex[resRecord->second.first.getID()] = newID;
+  mResources[newID] = resRecord->second;
+  mResources.erase(resRecord);
+
+  // Note down the cache hit and return true
+  mCacheHits++;
   return true;
 }
 
-void InMemoryResourceCache::dump(FILE* out) {
-  Block* first = last()->next;
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, (block == first) ? "┏━━━━━━━━━━━━━━━━" : "┳━━━━━━━━━━━━━━━━");
-  });
-  fprintf(out, "┓\n");
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, "┃ offset: %6zu ", block->offset);
-  });
-  fprintf(out, "┃\n");
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, "┃ size:   %6zu ", block->size);
-  });
-  fprintf(out, "┃\n");
-  foreach_block(first, [&](Block* block) {
-    if (block->isFree()) {
-      fprintf(out, "┃ free           ");
-    } else {
-      fprintf(out, "┃ id: %10.10s ", block->id.c_str());
-    }
-  });
-  fprintf(out, "┃\n");
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, (block == mHead) ? "┃ head           " : "┃                ");
-  });
-  fprintf(out, "┃\n");
-  foreach_block(first, [&](Block* block) {
-    fprintf(out, (block == first) ? "┗━━━━━━━━━━━━━━━━" : "┻━━━━━━━━━━━━━━━━");
-  });
-  fprintf(out, "┛\n");
+size_t InMemoryResourceCache::totalCacheSize() const { return mMemoryLimit; }
+
+size_t InMemoryResourceCache::unusedSize() const {
+  return (size_t)(mMemoryLimit - mMemoryUse);
 }
 
-bool InMemoryResourceCache::putCache(const Resource& resource,
-                                     const void* data) {
-  if (resource.size > mBufferSize) {
-    return false;  // Wouldn't fit even if everything was evicted.
+bool InMemoryResourceCache::resize(size_t newSize) {
+  // Throw things out of the cache until we're below limit.
+  if (newSize < mMemoryUse) {
+    assert(evictLeastRecentlyUsed(mMemoryUse - newSize));
   }
 
-  // Merge mHead into next block(s) until it is big enough to hold our resource.
-  while (mHead->size < resource.size) {
-    mHead->size += mHead->next->size;
-    destroy(mHead->next);
+  return true;
+}
+
+void InMemoryResourceCache::dump(FILE* file) { assert(false); }
+
+void InMemoryResourceCache::clear() {
+  for (auto&& resource : mResources) {
+    mAllocator->releaseAllocation(resource.second.second);
   }
 
-  if (mHead->size > resource.size) {
-    // We've got some left-over space in this block. Split it.
-    size_t space = mHead->size - resource.size;
-    size_t offset = (mHead->offset + resource.size) % mBufferSize;
-    auto next = new Block(offset, space);
-    next->linkAfter(mHead);
-    mHead->size = resource.size;
+  mResourceIndex.clear();
+  mResources.clear();
+
+  mMemoryUse = 0;
+}
+
+std::map<unsigned int, std::pair<Resource, MemoryAllocator::Handle> >::iterator
```

Comment:
std::map has awkward performance characteristics, but should be fine for the number of things we intend on putting into it.

Reply:
I've rarely had problems with them? Is there anything in particular you're worried about? Most things are O(logN) or less.

Reply:
Agreed offline to keep the map.